# Node Classification Practice - Tasks and Final Results

## Task 1
> Explain the metrics used in the report:
- Accuracy:
- Precision:
- Recall:
- F1 score:

> Why is it important to report all these metrics? Did you notice that the accuracy is not the best metric to report? Why?

---

<br>

La métrica "Accuracy" (exactitud) es la métrica mas sencilla y utilizada para la evaluación del rendimiento de un modelo de clasificación. Básicamente, indica qué tan preciso es el modelo en sus predicciones en terminos de porcentajes.

La fórmula para calcular la exactitud es:

$$ \text{Accuracy} = \frac{\text{Número de predicciones correctas}}{\text{Número total de predicciones}} \times 100\% $$

Como se ha explicado anteriormente, es facil de entender, de interpretar y buena como punto inicial para la evaluación de un modelo, aun asi cabe destacar las siguientes desventajas o limitaciones de la métrica "Accuracy":
- No considera el balanceo de las clases en el conjunto de datos. En presencia de clases desbalanceadas, puede dar una interpretación erronea de lo bien que se desenvuelve el clasificador.
- No discrimina entre diferentes tipos de errores de clasificación (falsos positivos vs. falsos negativos).

<br>

---

<br>

La métrica "Precision" (precisión) es una métrica que se utiliza en la clasificación multiclase para evaluar la exactitud de un modelo al identificar cada clase específica. Se calcula como la proporción de instancias que el modelo ha clasificado correctamente como pertenecientes a una clase, entre todas las instancias que el modelo ha identificado como pertenecientes a esa clase.

La fórmula para calcular la precisión aplicada a cada clase es:

$$ \text{Precision} = \frac{\text{True Positives}}{\text{True Positives} + \text{False Positives}} \ $$

La precisión es útil en la clasificación multiclase porque nos indica qué tan confiables son las predicciones para cada clase en particular. Cuanto más alta sea la precisión para una clase específica, menos instancias de otras clases serán incorrectamente clasificadas como pertenecientes a esa clase.

<br>

---

<br>

La métrica "Recall" (sensibilidad) es una métrica utilizada en la clasificación multiclase. Esta métrica indica qué tan bien un modelo puede identificar todas las instancias de una clase específica. Se calcula como la proporción de instancias positivas correctamente identificadas de una clase sobre todas las instancias que realmente son positivas de esa clase.

La fórmula para calcular el recall aplicado a cada clase es:

$$ \text{Recall} = \frac{\text{True Positives}}{\text{True Positives} + \text{False Negatives}} \ $$

El recall es útil porque indica qué tan bien el modelo puede encontrar todas las instancias positivas. Un valor de recall más alto indica que el modelo es más capaz de identificar correctamente las instancias positivas, lo que es importante en situaciones donde identificar todas las instancias positivas es crítico. En la clasificación multiclase, esto significa que el modelo es eficaz para identificar instancias de cada clase individual.

El recall se diferencia de la precisión en que, mientras la precisión se centra en la exactitud de las predicciones positivas del modelo, el recall se enfoca en la capacidad del modelo para identificar todas las instancias positivas. En otras palabras, la precisión mide la proporción de predicciones positivas que son correctas, mientras que el recall mide la proporción de instancias positivas reales que el modelo es capaz de identificar.

<br>

---

<br>

El F1 Score es una métrica que combina tanto la precisión como el recall en una sola medida, con el objetivo final de proporcionar una visión equilibrada del rendimiento de un modelo de clasificación. Se calcula como la media armónica de la precisión y el recall.

La fórmula para calcular el F1 Score es:

$$ \text{F1 Score} = 2 \times \frac{\text{Precision} \times \text{Recall}}{\text{Precision} + \text{Recall}} \ $$ 

El F1 Score es útil porque tiene en cuenta tanto la precisión como el recall, proporcionando una métrica única que refleja tanto la capacidad del modelo para clasificar correctamente las instancias positivas como su capacidad para encontrar todas las instancias positivas.

<br>

---

<br>

Es importante disponer de todas estas métricas porque cada una proporciona información diferente sobre el rendimiento del modelo. 

Por ejemplo, "Accuracy" puede ser una medida que de información erronea en cuanto al rendimiento del modelo  si las clases están desequilibradas. En este caso, un modelo que siempre predice la clase mayoritaria tendrá una alta exactitud, pero no será útil para predecir sobre la clase minoritaria.

La precisión y el recall son importantes y útiles en estos casos, ya que se centran en cómo el modelo se desempeña en la identificación de cada clase. 

Finalmente el F1 score proporciona una medida en la que se tiene en cuenta tanto la precisión como el recall.

Por lo tanto, aunque la exactitud puede ser una métrica útil, no siempre es la mejor para "juzgar" el modelo de clasificación, especialmente en casos de clases desequilibradas. Por eso es importante considerar y obtener los resultados de varias métricas al evaluar el rendimiento de un modelo de clasificación.

<br>

---


## TASK 3
> Make a summary of the architecture of the model you used. 
> 
> What are the parameters of the model? 
> 
> What is the number of parameters of the model? 
> 
> What is the number of trainable parameters of the model? Why they perform differently? 
> 
> What happens if you add to much layers of GNNs? 
> 
> During your summary, you can cite the different papers you used to implement your model.

### Resumen de la arquitectura.

Para la resolución de la siguiente practica se han creado una serie de redes neuronales para la clasificación de grafos, las redes neuronales creadas se basan en las siguientes estrategias:

- GATs (Graph Attention Networks)
  - Aprendizaje de representaciones de grafos con atención: utiliza mecanismos de atención para ponderar la importancia de los vecinos de un nodo en particular al calcular la representación del nodo.
- GraphSAGE (Graph Sample and Aggregate)
  - Generación de representaciones de nodos a partir de subgrafos y agregación: muestreo de subgrafos alrededor de los nodos y la agregación de información de los vecinos para generar representaciones de nodo.
- GINs (Graph Isomorphism Networks)
  - Aprendizaje de representaciones de grafos invariantes a la permutación de nodos: calcula características de cada nodo y agrega estas características para obtener la representación del grafo.

A continuación se explicarán las estructuras de los diferentes modelos de redes neuronales que se han creado para la resolución de la práctica.

#### Modelos Simples

Se trata de redes que siguen las estrategias explicadas anteriormente para la clasificación de nodos pero que mantienen la estructura básica de la red neuronal convolucional proporcionada como ejemplo.

La estructura de estas redes es la siguiente:
- Una primera capa que toma dataset.num_features (o la variable 'IN_CHANNELS', obtenida al inicio) como entrada y produce 'HIDDEN_CHANNELS' (variable creada al inicio también) como salida.
- Una segunda capa que toma hidden_channels como entrada y produce el numero de clases de cada conjunto de datos (variable dataset.num_classes) como salida.

Y la propagación de las redes se realiza de la siguiente manera:
- Una primera capa seguida de una función de activación ReLU.
- Una capa de dropout del 50%.
- Una segunda capa.
- Finalmente, se aplica una función de activación softmax y se devuelve la salida.

#### Modelos Complejos

Para analizar si los mismos modelos pero con una estructura mas compleja ofrecen mejores resultados que los modelos básicos, se han desarrollado nuevos modelos con mas capas.

Las características generalizadas y resumidas de estos modelos son las siguientes:
- Disponen de un mayor número de capas:
  - Por lo tanto, la propagación a lo largo de la red es más lenta.
- Dropout aplicado después de cada capa ReLU.

Concretamente, la estructura de estas redes es la siguiente:
- Una primera capa que toma in_channels características de entrada y produce hidden_channels características ocultas como salida.
- Capas intermedias que siguen el mismo patrón.
- Una última capa que toma hidden_channels características ocultas como entrada y produce out_channels características de salida.

Y la propagación de las redes se realiza de la siguiente manera:
- Cada capa está seguida de una función de activación, como ReLU.
- Entre capas, se aplica una capa de dropout con una tasa del 20%.
- La última capa se activa con una función de activación softmax y se devuelve la salida.


#### Modelos simples con capas espectrales

Los siguientes modelos que se han creado son identicos que los primeros modelos, el único cambio es la sustitución de la segunda capa por una capa espectral, la capa ChebConv.

Se trata de una capa de convolución que ha sido diseñada específicamente para trabajar grafos. Su estrategia es la de utilizar polinomios de Chebyshev para realizar la convolución en el dominio espectral de los grafos.

#### Modelos simples con capas espectrales y sin Softmax

Finalmente, los ultimos modelos desarrollados, y los que mejores resultados ofrecen, son los modelos anteriores, implementados con una capa espectral, a los que se le ha eliminado la función de activación softmax.

### ¿Cuáles son los parámetros del modelo? ¿Cuál es el número de parámetros del modelo? ¿Cuál es el número de parámetros entrenables del modelo? ¿Por qué funcionan de forma diferente? 

Los parámetros del modelo son los pesos y sesgos de las capas de las diferentes redes neuronales que se entrenan durante el desarrollo de la práctica, pero tambien existen parametros que se pueden definir manualmente. Los principales parametros ajustables que se repiten para la mayoria de modelos, asi como para sus diferentes estructuras, son los siguientes:
- hidden_channels: representa el número de canales ocultos o características en cada capa de la red neuronal, este parámetro establece la dimensionalidad de los vectores de características en cada capa, por lo que puede afectar la capacidad del modelo para aprender representaciones complejas de los datos.
- num_classes: Este parámetro indica el número de clases del problema de clasificación. Depende del dataset empleado, por ejemplo en el dataset Actor tiene un valor de 5. El objetivo de este parametro es el de determinar la dimensionalidad del espacio de salida de la red, lo que permite al modelo asignar una probabilidad a cada clase para una determinada entrada.
- num_heads: Este parámetro se utiliza solo en el modelo GAT para controlar el número de cabezas de atención en cada capa de atención múltiple. Esto permite que el modelo aprenda diferentes combinaciones lineales de información. Por lo que un mayor número de cabezas de atención puede aumentar la capacidad del modelo pero también aumenta la complejidad computacional y el tiempo de entrenamiento.
- p (p): Este parámetro se utiliza en la capa de dropout para controlar la probabilidad de que un elemento en el tensor de entrada se omita durante el entrenamiento. Es una técnica que ayuda a prevenir el sobreajuste al eliminar aleatoriamente un porcentaje de las unidades durante el entrenamiento. 

Por lo que la elección adecuada de los valores de los parametros es importante ya que puede influir en el rendimiento del modelo.

El número exacto de parámetros depende de las dimensiones de entrada y salida de cada capa, así como del número de cabezas en el caso de GAT. Por ejemplo, para el modelo MLP2 ajustado para predecir sobre el dataset Actor, se tienen los siguientes parametros
- Capa 1 (lin1): 932 (in_features) * 800 (out_features) + 800 (bias) = 746,400 parámetros.
- Capa 2 (lin2): 800 (in_features) * 320 (out_features) + 320 (bias) = 256,320 parámetros.
- Capa 3 (lin3): 320 (in_features) * 80 (out_features) + 80 (bias) = 25,680 parámetros.
- Capa 4 (lin4): 80 (in_features) * 5 (out_features) + 5 (bias) = 405 parámetros.

Lo que da un total de 746,400 + 256,320 + 25,680 + 405 = 1,028,805 parámetros.

En cuanto a los parametros entrenables, a excepcion del numero de clases (que depende exclusivamente del dataset), se pueden entrenar los modelos con diferentes combinaciones de los parametros ajustables para observar como influye en la exactitud de las predicciones. Al entrenar con diferentes valores, los parametros totales tambien varían ya que se actualizan durante el entrenamiento para reducir la funcion de perdida y mejorar las predicciones.

### ¿Qué ocurre si se añaden demasiadas capas de GNN? 

Agregar demasiadas capas a una red neuronal de clasificación de grafos puede provocar varios problemas:
- Sobreajuste: Aumentar el número de capas aumenta la capacidad del modelo, lo que puede provocar sobreajuste. Tener sobreajuste en el modelo implica que predice correctamente los datos de entrenamiento pero predice con muchos errores datos que no ha analizado todavia.
- Complejidad: Las redes más profundas requieren más cálculos, por lo que lleva a un aumento en el tiempo de entrenamiento y los recursos necesarios.
- Difusión de características (Feature Smoothing): En las GNN, cada capa agrega información de los nodos vecinos. Si hay demasiadas capas, las características de los nodos se mezclan tanto que la información original del nodo se pierde.

En mi experiencia durante el desarrollo de la practica, los modelos con más capas han empleado más tiempo de entrenamiento y han ofrecido peores resultados de clasificación. 

Resumiendo, al aumentar con un mayor número de capas la red neuronal, puede darse el caso del sobreajuste y a una propagación más lenta de la información a través de la red, que implica peores resultados de clasificación en el conjunto de validación.


## TASK 4
> During the experiments, you have used different datasets with different **Heterophily** and **Homophily**:

---

1.  **Explain what is the Heterophily and Homophily. Implement a function so that given a graph/dataset you obtain the edge-homophility.**


La Homofilia y la Hetereofilia son conceptos que en el contexto de los grafos y las redes, describen las tendencias y los patrones de conexion entre nodos.

La Homofilia se refiere al concepto en el que los nodos de una red o un grafo tienden a estar conectados con otros nodos que son similares a ellos, en cuanto a atributos o características.

Esto implica que los nodos que comparten características similares tienden a estar más interconectados entre sí que con nodos que tienen características diferentes, accion que tambien implica a la formación de "clusters".

Por el otro lado la Heterofilia es el contrario de la homofilia. Se refiere a los nodos en una red que tienden a conectarse con otros nodos que son diferentes a ellos en atributos o características.

La heterofilia puede fomentar la diversidad y la difusión de información entre diferentes grupos en la red, ya que los nodos con diferentes características pueden tener acceso a diferentes recursos y perspectivas.

En resumen, la homofilia en un grafo implica la formación de grupos o clusters similares, mientras que la heterofilia implica la diversidad y la conexión entre grupos con características diferentes.

---

A continuación se muestra el código de dos funciones que calculan la homofilia, hay dos estrategias: una para calcular la homofilia de las aristas y una funcion que calcula la homofilia del grafo.

In [2]:
import torch

def edge_homophily(data):
    edge_index = data.edge_index
    y = data.y

    edge_labels = y[edge_index]
    
    homophily_edges = (edge_labels[0] == edge_labels[1]).sum().item()

    total_edges = edge_index.size(1)

    edge_homophily = homophily_edges / total_edges

    return edge_homophily


def node_homophily(graph):
    total_homofilias = 0
    total_nodes = graph.num_nodes

    for v in range(total_nodes):
        neighbors = graph.edge_index[1][graph.edge_index[0] == v]
        same_label_neighbors = (graph.y[neighbors] == graph.y[v]).sum().item()
        degree_v = len(neighbors)
        if degree_v > 0:
            total_homofilias += same_label_neighbors / degree_v

    homofilia = total_homofilias / total_nodes if total_nodes > 0 else 0
    return homofilia

A continuación se muestra el grado de homofilia para cada grafo.

In [3]:
from torch_geometric.datasets import Planetoid, Actor, WikipediaNetwork, WebKB
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

dataset = Planetoid(root='/tmp/Cora', name='Cora') 
data_cora = dataset[0] 

dataset = Planetoid(root='/tmp/CiteSeer', name='CiteSeer') 
data_citeseer = dataset[0] 

dataset = Actor(root='/tmp/Actor') 
data_actor = dataset[0] 

dataset = WikipediaNetwork(root='/tmp/Chamaleon', name='Chameleon')
data_chamaleon = dataset[0]


dataset = WikipediaNetwork(root='/tmp/Squirrel', name='Squirrel')
data_squirrel = dataset[0] 

dataset = WebKB(root='/tmp/Wisconsin', name='Wisconsin') 
data_wisconsin = dataset[0]

dataset = WebKB(root='/tmp/Cornell', name='Cornell') 
data_cornell = dataset[0]

dataset = WebKB(root='/tmp/Texas', name='Texas') 
data_texas = dataset[0]

lista = [data_cora, data_citeseer, data_actor, data_chamaleon, data_squirrel, data_wisconsin, data_cornell, data_texas]
nombres = ['Cora', 'CiteSeer', 'Actor', 'Chamaleon', 'Squirrel', 'Wisconsin', 'Cornell', 'Texas']
resultados1 = []
resultados2 = []

for graph, nombre in zip(lista, nombres):
    homophily_score = edge_homophily(graph)
    resultados1.append({'Dataset': nombre, 'Edge Homophily': homophily_score})


for graph, nombre in zip(lista, nombres):
    homophily_score = node_homophily(graph)
    resultados2.append({'Dataset': nombre, 'Node Homophily': homophily_score})

df1 = pd.DataFrame(resultados1)
df2 = pd.DataFrame(resultados2)


df = df1.copy()
df['Node Homophily'] = df2['Node Homophily']
df

Processing...
Done!
Processing...
Done!
Processing...
Done!
Processing...
Done!
Processing...
Done!
Processing...
Done!
Processing...
Done!
Processing...
Done!


Dataset  Edge Homophily  Node Homophily
0       Cora        0.809966        0.825158
1   CiteSeer        0.735501        0.706249
2      Actor        0.218761        0.205491
3  Chamaleon        0.235007        0.249013
4   Squirrel        0.223943        0.219001
5  Wisconsin        0.196117        0.134107
6    Cornell        0.130872        0.106340
7      Texas        0.107692        0.103709

---
2.   **What is the impact of the Heterophily and Homophily on the performance of the model?**

La homofilia y la heterofilia tienen un impacto significativo en el rendimiento de los modelos de redes neuronales para la clasificación de grafos que se han creado durante el desarrollo de la práctica. 

La homofilia, como se ha explicado anteriormente, se refiere a la tendencia de los nodos similares a conectarse entre sí. En los datasets a clasificar, unicamente Cora y CiteSeer muestran altos niveles de homofilia tanto en las aristas como en los nodos. Esto significa que los nodos similares tienden a estar conectados, lo que facilita a priori la tarea de clasificación para los modelos, ya que la información de las etiquetas puede propagarse eficientemente a través de las conexiones.

En este caso, se observa como para la mayoria de los modelos creados, las tasas de acierto en la clasificación se mantienen bastante elevadas para los datasets Cora y CiteSeer, debido probablemente a lo explicado anteriormente.

Por otro lado, la heterofilia se refiere a la tendencia de los nodos diferentes a conectarse entre sí. La mayoria de datasets a clasificar presentan un bajo grado de homofilia (traducible a un alto grado de heterofilia), pero sobre todo destacan los datasets Cornell y Texas que muestran muy bajos niveles de homofilia. Esto puede hacer que la clasificación sea menos exacta para determinados modelos, ya que los nodos conectados tienen etiquetas diferentes, dificultando la propagación de la información de las etiquetas.

A pesar de lo expuesto en el apartado anterior, la mayoria de modelos creados ofrece buenos resultados de clasificación (con mas de un 70% de aciertos en el menor de los casos) en los datasets Cornell y Texas, por lo que se puede descartar el grado de homofilia como la caracteristica con mas influencia en los resultados de clasificación.

Estas son unas observaciones muy generalizadas, basandome únicamente en el grado de homofilia de los datasets y en los resultados de acierto de los modelos generales, por lo que pueden haber otros factores que también afecten al rendimiento del modelos.

---

---
3.   **Include the edge-homophily of each dataset in the table above. Explain your conclusions on behalf of the edge-homophily.**
---

In [9]:
df

Dataset  Edge Homophily  Node Homophily
0       Cora        0.809966        0.825158
1   CiteSeer        0.735501        0.706249
2      Actor        0.218761        0.205491
3  Chamaleon        0.235007        0.249013
4   Squirrel        0.223943        0.219001
5  Wisconsin        0.196117        0.134107
6    Cornell        0.130872        0.106340
7      Texas        0.107692        0.103709

---

La homofilia, aunque puede ser un indicador útil para analizar brevemente la estructura de los grafos, no es una variable con peso en la determinación de si una red neuronal de clasificación va a obtener buenos resultados de clasificación en el grafo.

Existen más factores que pueden influir en los resultados de los modelos de clasificación más allá de la homofilia, por ejemplo los siguientes:
- Complejidad de la red: La homofilia es solo una característica de la red. Hay muchas otras características estructurales que pueden influir en el rendimiento de los modelos, como la densidad de los grafos, el numero de caracteristicas, etc.
- Diversidad de los datos: Por ejemplo, si dos grafos tienen niveles similares de homofilia, si uno de ellos tiene una distribución de clases más equilibrada, puede ser más fácil de clasificar, o del modo contrario, si un grafo tiene muchas mas clases de un tipo que del otro, el modelo predecirá mayoritariamente a ese tipo de grafos, por lo que obtendrá buenos resultados en cuanto Accuracy pero no será un modelo fiable.
- Arquitectura del modelo: Diferentes modelos pueden tener diferentes sensibilidades a la homofilia debido a su arquitectura, a las capas y/o a la estrategia que utilicen, tal como se verá a continuación.
- Centralidad y Comunidades en los grafos: Estas carácteristicas son interesantes de estudiar ya que pueden proporcionar información importante sobre la estructura de los grafos, por lo que se podrá adaptar la arquitectura de los modelos a cada dataset en particular para realizar mejores predicciones.

Estos factores se pueden confirmar con los resultados de los modelos en los diferentes datasets, por ejemplo a continuación se van a comparar los datasets Wisconsin y Squirrel:
   - Wisconsin y Squirrel tienen niveles bajos de homofilia (0.196 y 0.224 respectivamente), los modelos de clasificación obtienen resultados muy diferentes en estos conjuntos de datos.
     - La mayor tasa de acierto en Squirrel la tiene el modelo MLP4, con un 42.23% de Accuracy
     - La mayor tasa de acierto en Wisconsin la tiene el modelo MLP4, con un 92.55% de Accuracy
   - La diferencia de acierto es de aproximadamente un 50% en el mejor de los casos, esto confimra que el grado de homofilia no influye en los resultados de clasificación ya que como se ha visto los modelos tienen un rendimiento significativamente mejor en Wisconsin en comparación con Squirrel.

Por el otro lado, si observamos el conjunto de datos de Cora, que tiene una alta homofilia, en sus resultados de clasificación todos los modelos muestran un rendimiento relativamente alto (GCN4 tiene un 82.21% y MLP4 un 83.01%). Por otro lado, el conjunto de datos de Texas tiene una baja homofilia y los modelos tienden a tener un rendimiento similar este conjunto de datos (GCN4 tiene un 80.54% y MLP4 un 92.97%).

Todos estos hechos sugieren que los factores explicados anteriormente pueden estar desempeñando un papel más importante en los resultados de clasificación que la homofilia.

En cuanto a los diferentes modelos empleados, estos son los modelos que mayor precision presentan en cada uno de los datasets:

| Dataset    | Modelo más preciso |
|------------|--------------------|
| Cora       | MLP4               |
| CiteSeer   | MLP4               |
| Actor      | MLP4               |
| Chamaleon  | MLP4               |
| Squirrel   | MLP4               |
| Wisconsin  | MLP4               |
| Cornell    | MLP4               |
| Texas      | MLP4               |

Como se observa el modelo MLP4, creado siguiendo la misma estructura que el modelo MLP proporcionado pro el profesor, pero sustituyendo las capas por espectrales y eliminando la función SoftMax al final, es el modelo que mejores resultados presenta en todos los datasets.

Dejando de lado los resultados de los modelos 4, se puede observar por ejemplo en los modelos básicos como algunos modelos son más sensibles a la homofilia que otros. Por ejemplo, MLP tiende a tener un rendimiento superior en datasets con baja homofilia (como Cornell o Texas) en comparación con el modelo GCN.

Esto puede ser debido a que el Perceptron Multicapa (MLP) no tinen en cuenta la estructura del grafo, por lo que el indice de homofilia es indiferente para este modelo ya que puede aprender patrones de características más simples sin tener en cuenta la relación entre nodos vecinos en el grafo.

En conclusión, el indice de homofilia puede ser útil para obtener una ligera idea de como se unen los nodos en los grafos, pero no es una caracteristica muy influyente en cuanto a la creación de mooedelos de clasificación de grafos.

---

## Final Results

A continuación se muestran los resultados finales, obtenidos en las ejecuciones de los modelos de redes neuronales en cada dataset.

### Basic Models


|  | Cora | Citeseer | Actor | Chamaleon | Squirrel | Wisconsin | Cornell | Texas |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| GCN | 80.63 $\pm{1.27}$ | 72.79 $\pm{1.32}$ | 30.75 $\pm{1.01}$ | 43.84 $\pm{1.66}$ | 34.38 $\pm{1.33}$ | 70.39 $\pm{6.41}$ | 64.86 $\pm{8.55}$ | 70.54 $\pm{5.33}$ |
| MLP | 74.54 $\pm{1.05}$ | 71.92 $\pm{1.60}$ | 35.59 $\pm{0.94}$ | 50.09 $\pm{1.79}$ | 37.20 $\pm{1.75}$ | 88.43 $\pm{6.47}$ | 84.59 $\pm{5.93}$ | 91.62 $\pm{4.90}$ |
| GAT | 75.97 $\pm{2.82}$ | 71.11 $\pm{0.86}$ | 30.59 $\pm{1.33}$ | 43.42 $\pm{1.62}$ | 33.76 $\pm{1.15}$ | 75.88 $\pm{5.89}$ | 70.54 $\pm{5.85}$ | 75.41 $\pm{4.59}$ |
| GraphSAGE | 78.01 $\pm{1.50}$ | 72.13 $\pm{1.31}$ | 35.48 $\pm{0.90}$ | 48.84 $\pm{1.23}$ | 37.51 $\pm{1.89}$ | 89.02 $\pm{6.40}$ | 80.54 $\pm{7.43}$ | 89.46 $\pm{7.40}$ |
| GIN | 79.85 $\pm{2.38}$ | 72.45 $\pm{1.78}$ | 32.10 $\pm{1.29}$ | 47.39 $\pm{4.52}$ | 35.57 $\pm{3.31}$ | 70.0 $\pm{8.46}$ | 72.43 $\pm{4.80}$ | 71.08 $\pm{5.41}$ |


### Complex Models

|      | Cora     | Citeseer | Actor   | Chamaleon | Squirrel | Wisconsin | Cornell  | Texas    |
|------|----------|----------|---------|-----------|----------|-----------|----------|----------|
| GCN2 | 76.72 $\pm{2.67}$ | 68.96 $\pm{2.08}$ | 28.28 $\pm{0.79}$ | 42.63 $\pm{1.54}$ | 34.57 $\pm{1.70}$ | 48.82 $\pm{8.79}$ | 37.29 $\pm{10.03}$ | 54.05 $\pm{13.07}$ |
| MLP2 | 72.92 $\pm{1.39}$ | 69.13 $\pm{1.69}$ | 36.06 $\pm{1.53}$ | 45.79 $\pm{1.68}$ | 34.61 $\pm{1.19}$ | 87.84 $\pm{6.55}$ | 83.24 $\pm{5.10}$ | 86.49 $\pm{8.11}$ |
| GAT2 | 31.11 $\pm{1.80}$ | 21.26 $\pm{5.12}$ | 21.54 $\pm{0.92}$ | 35.83 $\pm{7.13}$ | NO DATA | 46.08 $\pm{6.81}$ | 26.49 $\pm{12.43}$ | 54.86 $\pm{13.78}$ |
| GraphSAGE2 | 76.03 $\pm{1.86}$ | 69.20 $\pm{1.22}$ | 35.60 $\pm{1.20}$ | 47.00 $\pm{2.18}$ | 35.37 $\pm{1.77}$ | 73.92 $\pm{12.95}$ | 54.86 $\pm{15.24}$ | 65.14 $\pm{14.12}$ |
| GIN2 | 31.18 $\pm{12.57}$ | 28.23 $\pm{12.42}$ | 25.18 $\pm{0.65}$ | 40.24 $\pm{4.25}$ | 26.63 $\pm{2.41}$ | 46.67 $\pm{8.80}$ | 39.19 $\pm{16.85}$ | 49.73 $\pm{18.14}$ |


### Basic Models, Spectral Layers

|  | Cora | Citeseer | Actor | Chamaleon | Squirrel | Wisconsin | Cornell | Texas |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| GCN3 | 80.72 $\pm{1.25}$ | 73.14 $\pm{1.38}$ | 31.14 $\pm{0.58}$ | 46.25 $\pm{2.29}$ | 35.24 $\pm{1.99}$ | 78.82 $\pm{7.33}$ | 73.51 $\pm{9.11}$ | 76.76 $\pm{5.01}$ |
| MLP3 | 79.30 $\pm{1.24}$ | 72.94 $\pm{1.53}$ | 35.54 $\pm{1.09}$ | 51.43 $\pm{2.43}$ | 36.40 $\pm{2.24}$ | 88.04 $\pm{5.65}$ | 80.81 $\pm{6.45}$ | 85.68 $\pm{6.51}$ |
| GAT3 | 76.66 $\pm{3.31}$ | 71.55 $\pm{1.47}$ | 34.18 $\pm{0.64}$ | 44.63 $\pm{1.56}$ | 36.65 $\pm{2.02}$ | 80.59 $\pm{6.04}$ | 71.62 $\pm{4.87}$ | 78.92 $\pm{6.02}$ |
| GraphSAGE3 | 76.61 $\pm{1.99}$ | 72.75 $\pm{1.49}$ | 34.77 $\pm{0.70}$ | 50.55 $\pm{2.22}$ | 35.80 $\pm{1.58}$ | 88.43 $\pm{5.78}$ | 81.62 $\pm{9.80}$ | 83.24 $\pm{7.63}$ |
| GIN3 | 79.39 $\pm{1.78}$ | 71.61 $\pm{1.32}$ | 32.77 $\pm{1.71}$ | 48.03 $\pm{2.87}$ | 34.41 $\pm{1.99}$ | 80.00 $\pm{7.73}$ | 74.59 $\pm{7.47}$ | 78.65 $\pm{5.85}$ |


### Basic Models, Spectral Layers, without SoftMax

|  | Cora | Citeseer | Actor | Chamaleon | Squirrel | Wisconsin | Cornell | Texas |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| GCN4 | 82.21 $\pm{1.13}$ | 74.79 $\pm{2.14}$ | 33.04 $\pm{1.26}$ | 51.86 $\pm{4.49}$ | 40.53 $\pm{4.74}$ | 78.04 $\pm{6.25}$ | 72.70 $\pm{12.23}$ | 80.54 $\pm{6.49}$ |
| MLP4 | 83.01 $\pm{1.51}$ | 76.70 $\pm{2.00}$ | 37.78 $\pm{2.09}$ | 59.25 $\pm{4.46}$ | 42.23 $\pm{4.89}$ | 92.55 $\pm{7.43}$ | 86.49 $\pm{8.55}$ | 92.97 $\pm{5.16}$ |
| GAT4 | 79.11 $\pm{2.40}$ | 72.93 $\pm{2.02}$ | 36.74 $\pm{2.10}$ | 46.38 $\pm{2.32}$ | 39.86 $\pm{2.16}$ | 84.31 $\pm{7.39}$ | 78.92 $\pm{7.33}$ | 83.51 $\pm{6.56}$ |
| GraphSAGE4 | 80.46 $\pm{1.86}$ | 74.91 $\pm{1.56}$ | 37.64 $\pm{1.69}$ | 58.07 $\pm{4.61}$ | 41.41 $\pm{4.23}$ | 90.59 $\pm{6.55}$ | 83.24 $\pm{10.66}$ | 89.19 $\pm{6.04}$ |
| GIN4 | 80.48 $\pm{2.62}$ | 75.80 $\pm{1.95}$ | 34.22 $\pm{1.77}$ | 48.16 $\pm{5.75}$ | 33.84 $\pm{5.73}$ | 82.35 $\pm{9.40}$ | 73.78 $\pm{13.19}$ | 84.59 $\pm{4.84}$ |